File: src/main/java/gargoyle/calendar/web/data/ExceptionInfo.java
package gargoyle.calendar.web.data;

import gargoyle.calendar.util.Interpolation;
import lombok.Value;

@Value
public class ExceptionInfo {
    private static final String TAIL = "Exception";

    String message;
    String details;

    public ExceptionInfo(Exception exception) {
        message = Interpolation.splitWords(Interpolation.cutTail(exception.getClass().getSimpleName(), TAIL),
                Interpolation.SPLIT_WORDS_OPTIONS.FIRST_UPPER);
        details = Interpolation.findFirstNotEmpty(new ThrowableIterable(exception), Throwable::getLocalizedMessage);
    }

}

File: cal-web/src/main/kotlin/gargoyle/calendar/web/data/ExceptionInfo.kt
package gargoyle.calendar.web.data

data class ExceptionInfo(val message: String, val details: String) {
    constructor(exception: Throwable) : this(
        message =
        splitWords(exception::class.simpleName?.substringBeforeLast(TAIL) ?: TAIL, SplitWordsOptions.FIRST_UPPER),
        details =
        generateSequence(exception) { it.cause }.map { it.localizedMessage ?: "" }.firstOrNull { it.isNotBlank() } ?: ""
    )

    companion object {
        private const val TAIL = "Exception"
        fun splitWords(text: String, options: SplitWordsOptions): String {
            var result = ""
            var word: Boolean
            text.forEachIndexed { i, c ->
                if (!c.isLetterOrDigit()) {
                    if (i != 0) result += ' '
                    return@forEachIndexed
                } else {
                    if (c.isUpperCase()) {
                        word = true
                        if (i != 0) result += ' '
                    } else {
                        word = false
                    }
                }
                result += when (options) {
                    SplitWordsOptions.FIRST_UPPER_EVERY -> if (word) c.uppercaseChar() else c.lowercaseChar()
                    SplitWordsOptions.FIRST_UPPER -> if (word && i == 0) c.uppercaseChar() else c.lowercaseChar()
                    SplitWordsOptions.ALL_LOWER -> c.lowercaseChar()
                    SplitWordsOptions.ALL_UPPER -> c.uppercaseChar()
                }
            }
            return result
        }

        enum class SplitWordsOptions {
            ALL_UPPER, ALL_LOWER, FIRST_UPPER, FIRST_UPPER_EVERY
        }
    }
}
